pkg/apis/pipeline/v1alpha1/taskrun_types.go

```diff
@@ -125,10 +124,9 @@ type TaskRun struct {
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	// +optional
-	Spec TaskRunSpec `json:"spec,omitempty"`
+	Spec *TaskRunSpec `json:"spec,omitempty"`
 	// +optional
-	//TODO(aaron-prindle) change back to TaskRunStatus
-	Status buildv1alpha1.BuildStatus `json:"status,omitempty"`
+	Status *TaskRunStatus `json:"status,omitempty"`
```

Comment:
Also why pointer, in this example in [serving](https://github.com/knative/serving/blob/1fa7f094427e2e519947475069915ecb0e8837fb/pkg/apis/autoscaling/v1alpha1/kpa_types.go#L41) its not used as a pointer to make sure its initialized to a value. Otherwise there will be no initial status and other checks could blow up

Reply:
Interesting! that's a good point, I'll try to dig a bit more and see what's going on

Reply:
I think I got a bit confused about json + nullable fields and conflated the problems:

1. If you mark a field as optional but it isn't nullable, it will still be present in generated json, but all the fields will be empty <-- this was unrelated to the problem I was seeing after all! :sweat_smile: 
2. If the field is required but not present, it will be populated with default values; if the values inside this required field are not optional, then there is a problem b/c they have to be present, so in [the serving example you linked](https://github.com/knative/serving/blob/1fa7f094427e2e519947475069915ecb0e8837fb/pkg/apis/autoscaling/v1alpha1/kpa_types.go#L119) the `Status` fields are all optional, same with [`Route`](https://github.com/knative/serving/blob/1fa7f094427e2e519947475069915ecb0e8837fb/pkg/apis/serving/v1alpha1/route_types.go#L127)

I think the real problem was that the fields in `Status` were required, so I'll make those optional instead.

Thanks for following up on this @pivotal-nader-ziada !!

Reply:
Thanks for fixing it

Reply:
Looking into it a bit more it seems the situation is actually slightly worse!

`knative/serving` [isn't using openAPIV3Schema validation](https://github.com/knative/serving/blob/1fa7f094427e2e519947475069915ecb0e8837fb/config/300-route.yaml) but [we are](https://github.com/knative/build-pipeline/blob/462cbd64fbeb09f723abc3317d1c445d2f15f462/config/300-taskrun.yaml#L15) ... it looks like this is why I'm running into errors even though this field is not required: https://github.com/kubernetes/apiextensions-apiserver/issues/25

For now the best workaround I can think of is to make these pointers so they can be null - but I think I should open an issue to investigate this further :S

(Open to other ideas too!)

Reply:
I think if serving isn't using openapi validation, we shouldn't either (build doesn't either). ISTR there were issues with getting it to work with our controllers, possible exactly the issues you're encountering now.

Reply:
If it's annotation-based validation you're after, you might want https://github.com/knative/pkg/pull/37

Personally I tend to prefer just writing code myself with tests, but 🤷‍♂️

Reply:
consistency across Knative repos is much better